File: spark-atomistic/src/spark_atomistic/rng.py
# Independently authored from OFFLATTICE_OTF_KMC_SPEC_V1 + ERRATA_1 + ERRATA_2.
# Base: 8e21e556838d997aa257db1a194c3ed88018ad3d77dcf1d811a82f7194f49c84
# Erratum 1: 52c9c8bc9e6839ec04709da99aa7cca39963f4efd6fc420a865d6bb0b911ec40
# Erratum 2: eecc5f13c72d90ae2b5c6ff99385cb51f69c0d0500b563eb3a8df2e7d9863995
"""Errata-2 Philox state, derivation, counter, and midpoint52 mapping."""

from __future__ import annotations

from dataclasses import dataclass
import hashlib
import struct
from typing import Any

from .errors import DomainFailure


MASK32 = 0xFFFFFFFF
MASK128 = (1 << 128) - 1
M0 = 0xD2511F53
M1 = 0xCD9E8D57
W0 = 0x9E3779B9
W1 = 0xBB67AE85
ALGORITHM = "Philox4x32-10:errata-1-midpoint52"


def _mul_hi_lo(a: int, b: int) -> tuple[int, int]:
    product = a * b
    return (product >> 32) & MASK32, product & MASK32


def philox4x32_10(counter: tuple[int, int, int, int],
                   key: tuple[int, int]) -> tuple[int, int, int, int]:
    c0, c1, c2, c3 = counter
    k0, k1 = key
    for _ in range(10):
        hi0, lo0 = _mul_hi_lo(M0, c0)
        hi1, lo1 = _mul_hi_lo(M1, c2)
        c0, c1, c2, c3 = ((hi1 ^ c1 ^ k0) & MASK32, lo1,
                          (hi0 ^ c3 ^ k1) & MASK32, lo0)
        k0 = (k0 + W0) & MASK32
        k1 = (k1 + W1) & MASK32
    return c0, c1, c2, c3


def words_to_uniform(a: int, b: int) -> float:
    if not (0 <= a <= MASK32 and 0 <= b <= MASK32):
        raise DomainFailure("INVALID_STATE", "state invalid", component="rng",
                            requirement="E1-DET-002-B")
    q = (a << 20) | (b >> 12)
    return (2 * q + 1) * (2.0 ** -53)


def _words_to_int(words: tuple[int, int, int, int] | list[int]) -> int:
    return sum(word << (32 * index) for index, word in enumerate(words))


def _int_to_words(value: int) -> tuple[int, int, int, int]:
    return tuple((value >> (32 * index)) & MASK32 for index in range(4))  # type: ignore[return-value]


@dataclass(slots=True)
class PhiloxStream:
    key: tuple[int, int]
    initial_counter: int = 0
    next_counter: int = 0
    consumed_blocks: int = 0
    consumed_uniforms: int = 0
    buffered_block: tuple[int, int, int, int] | None = None
    next_pair: int = 0

    def clone(self) -> "PhiloxStream":
        return PhiloxStream(self.key, self.initial_counter, self.next_counter,
                            self.consumed_blocks, self.consumed_uniforms,
                            self.buffered_block, self.next_pair)

    def commit_from(self, clone: "PhiloxStream") -> None:
        if clone.key != self.key or clone.initial_counter != self.initial_counter:
            raise DomainFailure("INVALID_STATE", "state invalid", component="rng",
                                requirement="E2-KMC-003")
        self.next_counter = clone.next_counter
        self.consumed_blocks = clone.consumed_blocks
        self.consumed_uniforms = clone.consumed_uniforms
        self.buffered_block = clone.buffered_block
        self.next_pair = clone.next_pair

    def uniform(self) -> float:
        if self.buffered_block is None:
            self.buffered_block = philox4x32_10(_int_to_words(self.next_counter), self.key)
            self.next_counter = (self.next_counter + 1) & MASK128
            self.consumed_blocks += 1
            self.next_pair = 0
        offset = 2 * self.next_pair
        result = words_to_uniform(self.buffered_block[offset], self.buffered_block[offset + 1])
        self.consumed_uniforms += 1
        if self.next_pair == 0:
            self.next_pair = 1
        else:
            self.buffered_block = None
            self.next_pair = 0
        return result

    def checkpoint(self) -> dict[str, Any]:
        return {"algorithm": ALGORITHM,
                "buffered_block": None if self.buffered_block is None else list(self.buffered_block),
                "consumed_blocks": self.consumed_blocks,
                "consumed_uniforms": self.consumed_uniforms,
                "initial_counter": list(_int_to_words(self.initial_counter)),
                "key": list(self.key),
                "next_counter": list(_int_to_words(self.next_counter)),
                "next_pair": self.next_pair}

    @classmethod
    def restore(cls, value: Any) -> "PhiloxStream":
        required = {"algorithm", "buffered_block", "consumed_blocks", "consumed_uniforms",
                    "initial_counter", "key", "next_counter", "next_pair"}
        if not isinstance(value, dict) or set(value) != required:
            raise DomainFailure("CHECKPOINT_CORRUPT", "checkpoint corrupt", component="checkpoint",
                                requirement="E2-RNG-002")
        key = value["key"]
        initial = value["initial_counter"]
        following = value["next_counter"]
        buffered = value["buffered_block"]
        arrays = [key, initial, following] + ([] if buffered is None else [buffered])
        if (value["algorithm"] != ALGORITHM or not isinstance(key, list) or len(key) != 2
                or any(not isinstance(item, list) or len(item) != 4 for item in arrays[1:])
                or any(type(word) is not int or not 0 <= word <= MASK32
                       for item in arrays for word in item)
                or type(value["consumed_blocks"]) is not int or value["consumed_blocks"] < 0
                or type(value["consumed_uniforms"]) is not int or value["consumed_uniforms"] < 0
                or value["next_pair"] not in {0, 1}):
            raise DomainFailure("CHECKPOINT_CORRUPT", "checkpoint corrupt", component="checkpoint",
                                requirement="E2-RNG-002")
        initial_integer = _words_to_int(initial)
        next_integer = _words_to_int(following)
        blocks = value["consumed_blocks"]
        uniforms = value["consumed_uniforms"]
        expected_buffered = uniforms % 2 == 1
        if (blocks != (uniforms + 1) // 2
                or next_integer != (initial_integer + blocks) & MASK128
                or (buffered is not None) != expected_buffered
                or value["next_pair"] != (1 if expected_buffered else 0)):
            raise DomainFailure("CHECKPOINT_CORRUPT", "checkpoint corrupt", component="checkpoint",
                                requirement="E2-RNG-002")
        if buffered is not None:
            previous = (next_integer - 1) & MASK128
            if tuple(buffered) != philox4x32_10(_int_to_words(previous), (key[0], key[1])):
                raise DomainFailure("CHECKPOINT_CORRUPT", "checkpoint corrupt", component="checkpoint",
                                    requirement="E2-RNG-002")
        return cls((key[0], key[1]), initial_integer, next_integer, blocks, uniforms,
                   None if buffered is None else tuple(buffered), value["next_pair"])


def _stream_from_digest(material: bytes) -> PhiloxStream:
    output = hashlib.sha256(material).digest()
    key = struct.unpack(">II", output[:8])
    counter_words = struct.unpack(">IIII", output[8:24])
    counter = _words_to_int(list(counter_words))
    return PhiloxStream(key, counter, counter)


def derive_trajectory_stream(seed: int) -> PhiloxStream:
    return _stream_from_digest(b"spark-trajectory-stream/2\0" + struct.pack(">Q", seed))


def derive_saddle_stream(seed: int, state_id: str, search_class: str,
                         search_index: int) -> PhiloxStream:
    state_bytes = state_id.encode("utf-8")
    class_bytes = search_class.encode("utf-8")
    material = (b"spark-saddle-substream/2\0" + struct.pack(">Q", seed)
                + struct.pack(">I", len(state_bytes)) + state_bytes
                + struct.pack(">I", len(class_bytes)) + class_bytes
                + struct.pack(">Q", search_index))
    return _stream_from_digest(material)


def derive_stream(seed: int, state_id: str, search_class: str,
                  search_index: int) -> PhiloxStream:
    if state_id == "trajectory" and search_class == "trajectory":
        return derive_trajectory_stream(seed)
    return derive_saddle_stream(seed, state_id, search_class, search_index)

File: spark-atomistic-rs/src/rng.rs
// Clean-room implementation from OFFLATTICE_OTF_KMC_SPEC_V1 + ERRATA_1.
// Independently authored; no implementation source consulted.
use crate::identity::hex_sha256;
use crate::status::{Status, StatusCode};
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use std::collections::BTreeMap;

pub const RNG_ALGORITHM: &str = "Philox4x32-10:errata-1-midpoint52";
const M0:u32=0xD251_1F53; const M1:u32=0xCD9E_8D57;
const W0:u32=0x9E37_79B9; const W1:u32=0xBB67_AE85;

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct PhiloxState {
    pub algorithm:String,
    pub key:[u32;2],
    pub initial_counter:[u32;4],
    pub next_counter:[u32;4],
    pub buffered_block:Option<[u32;4]>,
    pub next_pair:u8,
    pub consumed_blocks:u64,
    pub consumed_uniforms:u64,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct Philox { state:PhiloxState }

impl Philox {
    pub fn new(key:[u32;2],counter:[u32;4])->Self{Self{state:PhiloxState{algorithm:RNG_ALGORITHM.to_owned(),key,initial_counter:counter,next_counter:counter,buffered_block:None,next_pair:0,consumed_blocks:0,consumed_uniforms:0}}}
    pub fn from_state(state:PhiloxState)->Result<Self,Status>{validate_state(&state)?;Ok(Self{state})}
    pub fn state(&self)->PhiloxState{self.state.clone()}

    pub fn next_uniform(&mut self)->Result<f64,Status>{
        if self.state.buffered_block.is_none(){
            let block=philox4x32_10(self.state.next_counter,self.state.key);
            increment_counter(&mut self.state.next_counter)?;
            self.state.consumed_blocks=self.state.consumed_blocks.checked_add(1).ok_or_else(||rng_limit("consumed-block counter overflow"))?;
            self.state.buffered_block=Some(block);self.state.next_pair=0;
        }
        let block=self.state.buffered_block.ok_or_else(||rng_corrupt("missing buffered block"))?;
        let (a,b)=match self.state.next_pair{0=>(block[0],block[1]),1=>(block[2],block[3]),_=>return Err(rng_corrupt("invalid pair index"))};
        self.state.consumed_uniforms=self.state.consumed_uniforms.checked_add(1).ok_or_else(||rng_limit("uniform counter overflow"))?;
        self.state.next_pair+=1;
        if self.state.next_pair==2{self.state.buffered_block=None;self.state.next_pair=0;}
        uniform_from_words(a,b)
    }

    pub fn two_uniforms_atomic(&self)->Result<(f64,f64,Philox),Status>{
        let mut next=self.clone();let a=next.next_uniform()?;let b=next.next_uniform()?;Ok((a,b,next))
    }
}

pub fn uniform_from_words(a:u32,b:u32)->Result<f64,Status>{let q=((a as u64)<<20)|((b as u64)>>12);let u=((2*q+1) as f64)*2f64.powi(-53);if !(u>0.0&&u<1.0){return Err(rng_corrupt("Errata-1 midpoint escaped open interval"));}Ok(u)}

pub fn validate_state(s:&PhiloxState)->Result<(),Status>{
    if s.algorithm!=RNG_ALGORITHM{return Err(rng_corrupt("RNG algorithm mismatch"));}
    let expected_blocks=s.consumed_uniforms.checked_add(1).ok_or_else(||rng_corrupt("uniform count overflow"))?/2;
    if s.consumed_blocks!=expected_blocks{return Err(rng_corrupt("consumed block/uniform relationship invalid"));}
    if add_blocks(s.initial_counter,s.consumed_blocks)?!=s.next_counter{return Err(rng_corrupt("persistent Philox counter does not equal initial counter plus consumed blocks"));}
    let odd=s.consumed_uniforms%2==1;
    if odd!=(s.buffered_block.is_some()) || (odd && s.next_pair!=1) || (!odd && s.next_pair!=0){
        return Err(rng_corrupt("counter/buffer/consumed relationship invalid"));
    }
    if odd{let previous=subtract_one(s.next_counter)?;if s.buffered_block!=Some(philox4x32_10(previous,s.key)){return Err(rng_corrupt("buffered words are not the actual previous Philox block"));}}
    Ok(())
}

pub fn derive_substream(run_seed:&[u8],state_id:&str,search_class:&str,search_index:u64)->Philox{
    let mut h=Sha256::new();h.update(b"spark-saddle-substream/1\0");h.update(run_seed);h.update([0]);h.update(state_id.as_bytes());h.update([0]);h.update(search_class.as_bytes());h.update(search_index.to_be_bytes());
    let d=h.finalize();
    let word=|i:usize|u32::from_be_bytes([d[i],d[i+1],d[i+2],d[i+3]]);
    Philox::new([word(0),word(4)],[word(8),word(12),word(16),word(20)])
}

pub fn derive_trajectory_stream(run_seed:u64)->Philox{derive_e2(b"spark-trajectory-stream/2\0",run_seed,&[])}
pub fn derive_saddle_substream(run_seed:u64,state_id:&str,search_class:&str,search_index:u64)->Result<Philox,Status>{let state_len:u32=state_id.len().try_into().map_err(|_|rng_limit("state ID exceeds len32"))?;let class_len:u32=search_class.len().try_into().map_err(|_|rng_limit("search class exceeds len32"))?;let mut rest=Vec::new();rest.extend(state_len.to_be_bytes());rest.extend(state_id.as_bytes());rest.extend(class_len.to_be_bytes());rest.extend(search_class.as_bytes());rest.extend(search_index.to_be_bytes());Ok(derive_e2(b"spark-saddle-substream/2\0",run_seed,&rest))}
fn derive_e2(label:&[u8],run_seed:u64,rest:&[u8])->Philox{let mut h=Sha256::new();h.update(label);h.update(run_seed.to_be_bytes());h.update(rest);let d=h.finalize();let word=|i:usize|u32::from_be_bytes([d[i],d[i+1],d[i+2],d[i+3]]);Philox::new([word(0),word(4)],[word(8),word(12),word(16),word(20)])}

pub fn substream_digest(p:&Philox)->String{
    let s=p.state();let bytes=crate::checkpoint::canonical_json_bytes(&s).expect("validated finite integer RNG state serializes");format!("sha256:{}",hex_sha256(&bytes))
}

pub fn philox4x32_10(mut c:[u32;4],mut k:[u32;2])->[u32;4]{
    for round in 0..10{
        let p0=(M0 as u64)*(c[0] as u64);let p1=(M1 as u64)*(c[2] as u64);
        c=[(p1>>32) as u32^c[1]^k[0],p1 as u32,(p0>>32) as u32^c[3]^k[1],p0 as u32];
        if round!=9{k[0]=k[0].wrapping_add(W0);k[1]=k[1].wrapping_add(W1);}
    }c
}

fn increment_counter(c:&mut[u32;4])->Result<(),Status>{for x in c.iter_mut(){let(v,carry)=x.overflowing_add(1);*x=v;if !carry{return Ok(());}}Err(rng_limit("Philox counter exhausted"))}
fn subtract_one(mut c:[u32;4])->Result<[u32;4],Status>{for x in &mut c{let(v,borrow)=x.overflowing_sub(1);*x=v;if !borrow{return Ok(c)}}Err(rng_corrupt("buffered block precedes zero counter"))}
fn add_blocks(c:[u32;4],blocks:u64)->Result<[u32;4],Status>{let value=(c[0]as u128)|((c[1]as u128)<<32)|((c[2]as u128)<<64)|((c[3]as u128)<<96);let out=value.checked_add(blocks as u128).ok_or_else(||rng_corrupt("persistent counter addition overflow"))?;Ok([out as u32,(out>>32)as u32,(out>>64)as u32,(out>>96)as u32])}
fn rng_corrupt(m:&str)->Status{Status::simple(StatusCode::CheckpointCorrupt,"rng","DET-001",m)}
fn rng_limit(m:&str)->Status{Status::simple(StatusCode::ResourceLimit,"rng","DET-001",m)}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct RngRecord { pub trajectory:PhiloxState, pub substream_map:BTreeMap<String,String> }
